crates/polars-stream/src/nodes/dynamic_group_by.rs

```diff
@@ -0,0 +1,485 @@
+        ensure_duration_matches_dtype(period, index_dtype, "period")?;
+        ensure_duration_matches_dtype(offset, index_dtype, "offset")?;
+
+        use DataType as DT;
+        let (tu, tz) = match index_dtype {
+            DT::Datetime(tu, tz) => (*tu, tz.clone()),
+            DT::Date => (TimeUnit::Microseconds, None),
+            DT::Int64 | DT::Int32 => (TimeUnit::Nanoseconds, None),
+            dt => polars_bail!(
+                ComputeError:
+                "expected any of the following dtypes: {{ Date, Datetime, Int32, Int64 }}, got {}",
+                dt
+            ),
+        };
+
+        let buf_df = DataFrame::empty_with_arc_schema(schema.clone());
+        let buf_index_column =
+            Column::new_empty(index_column.clone(), &DT::Datetime(tu, tz.clone()));
+
+        // @NOTE: This is a bit strange since it ignores errors, but it mirrors the in-memory
+        // engine.
+        let tz = tz.and_then(|tz| tz.parse::<Tz>().ok());
+        let windower = GroupByDynamicWindower::new(
+            period,
+            offset,
+            every,
+            start_by,
+            closed_window,
+            tu,
+            tz,
+            include_boundaries || matches!(label, Label::Left),
+            include_boundaries || matches!(label, Label::Right),
+        );
+
+        let (slice_offset, slice_length) = slice.unwrap_or((0, IdxSize::MAX));
+
+        Ok(Self {
+            buf_df,
+
+            buf_df_offset: 0,
+            buf_index_column,
+            seq: MorselSeq::default(),
+
+            slice_offset,
+            slice_length,
+
+            group_by: None,
+            index_column,
+            index_column_idx,
+            label,
+            include_boundaries,
+            windower,
+            aggs,
+        })
+    }
+
+    #[expect(clippy::too_many_arguments)]
+    async fn evaluate_one(
+        windows: Vec<[IdxSize; 2]>,
+        lower_bound: Vec<i64>,
+        upper_bound: Vec<i64>,
+        aggs: &[(PlSmallStr, StreamExpr)],
+        state: &ExecutionState,
+        mut df: DataFrame,
+
+        group_by: Option<&str>,
+        index_column_name: &str,
+        index_column_idx: usize,
+        label: Label,
+        include_boundaries: bool,
+    ) -> PolarsResult<DataFrame> {
+        let height = windows.len();
+        let groups = GroupsType::Slice {
+            groups: windows,
+            overlapping: true,
+        }
+        .into_sliceable();
+
+        // @NOTE:
+        // Rechunk so we can use specialized rolling/dynamic kernels.
+        df.rechunk_mut();
+
+        let mut columns =
+            Vec::with_capacity(if include_boundaries { 2 } else { 0 } + 1 + aggs.len());
+
+        // Construct `lower_bound`, `upper_bound` and `key` columns that might be included in the
+        // output dataframe.
+        {
+            let mut lower = Int64Chunked::new_vec(PlSmallStr::from_static(LB_NAME), lower_bound);
```

Comment:
I feel like in a lot of cases we end up creating/casting/casting again the lower/upper when we don't need it at all. E.g. `Label::DataPoint` without `include_boundaries`.